Add an alternative method for corridor delineation

SarahAlidoost opened the pull request:
closes #8

This PR includes the lateral distance of edges (sum of its node distances) from the initial corridor guess in defining the shortest path.

SarahAlidoost commented:
@fnattino I couldn't apply the approach to the entire areas instead of `areas[1]` and `areas[2]`, due to two reasons:

1. The `trim` function drops nodes within the `corridor_initial`, which causes the `filter(group_components() == 1)` command to disconnect the upper and lower parts of the network. If we don't drop these nodes, the shortest paths will go inside the corridor.
2. The `st_network_paths` function requires specifying `'from'` and `'to'` points. When using the entire network, there are four target points. Still, we need to run the `st_network_paths` function separately for the upper and lower parts of the network. Otherwise, we cannot create a buffer (region) around the river.

fnattino commented on notebooks/corridor/corridor_bucharest.qmd:
```diff
@@ -413,7 +423,107 @@ st_write(
+        mutate(edge_length = edge_length())
+}
+
+# add sum of distances to each edge
```
This is a nice way to get some sort of distance between an edge and the region, by using the sum of the distances to the start- and end- nodes of the edge. 

Maybe even simpler, could one use the distance computed directly from an edge to the corridor? I have tried to run the following:
```
edges <- network |> activate("edges") |> st_geometry()
dists <- st_distance(edges, corridor_initial, which = "Euclidean") 
```
which **I think** returns the shortest distance between each edge and the corridor. Do you think this could work?

SarahAlidoost replied:
Yes, you are right. in this case, `st_distance` selects the minimum value. I was thinking about a case like below, which edge (1 or 2) is selected considering only one node (the nearest to the initial corridor)? I did a test for this notebook. The final results were the same. So, I replaced the sum of the distances of nodes with your suggestions. It makes the code simpler. 
![network_dist](https://github.com/user-attachments/assets/8551d0ec-705f-46f4-b145-d9ea6803c013)

fnattino reviewed (commented):
Hi @SarahAlidoost, thanks a lot for the great work, it's very promising! I have left a couple of questions/comments it would be great to hear your opinion on these points.

Concerning the points your raised on not being able to run the algorithm on the full network:

> 1. The trim function drops nodes within the corridor_initial, which causes the filter(group_components() == 1) command to disconnect the upper and lower parts of the network. If we don't drop these nodes, the shortest paths will go inside the corridor.

I think it is actually not a problem if the shortest paths slightly enter the corridor (ultimately, this initial corridor shape should represent just a first "guess" of where the corridor should lie), as long as one does not fully cross it to end up on the other side of the corridor. 

> 2. The st_network_paths function requires specifying 'from' and 'to' points. When using the entire network, there are four target points. Still, we need to run the st_network_paths function separately for the upper and lower parts of the network. Otherwise, we cannot create a buffer (region) around the river.

Yes, we would still need to run the shortest path algorithm twice, once for each side of the river. What I was thinking was:
* split the initial corridor buffer in two lines, which should represent the initial guesses of the corridor edges on the two sides
* find the start/end points for each of the side of the river by intersecting the two lines with the `bbox`
* run the method you have implemented using the full network for each pair of points and using the corresponding line as target geometry (instead of the polygon representing the initial guess of the corridor).

What do you think? Not needed to be addressed here/now, but I would love to hear your opinion on this.

SarahAlidoost commented on notebooks/corridor/corridor_bucharest.qmd:
```diff
@@ -413,7 +423,107 @@ st_write(
+        mutate(
+            edge_length = set_units(edge_length, "m"),
+            nodes_dist_sum = set_units(nodes_dist_sum, "m"),
+            weight = edge_length + nodes_dist_sum
```
Added.

SarahAlidoost commented on notebooks/corridor/corridor_bucharest.qmd:
````diff
@@ -413,7 +423,107 @@ st_write(
+
+```{r}
+# add id to nodes
+add_node_id <- function(net){
````
removed.

SarahAlidoost commented:
> Hi @SarahAlidoost, thanks a lot for the great work, it's very promising! I have left a couple of questions/comments it would be great to hear your opinion on these points.

Thanks @fnattino . I addressed your suggestions. lease see below my replies.

> Concerning the points your raised on not being able to run the algorithm on the full network:
> 
> > 1. The trim function drops nodes within the corridor_initial, which causes the filter(group_components() == 1) command to disconnect the upper and lower parts of the network. If we don't drop these nodes, the shortest paths will go inside the corridor.
> 
> I think it is actually not a problem if the shortest paths slightly enter the corridor (ultimately, this initial corridor shape should represent just a first "guess" of where the corridor should lie), as long as one does not fully cross it to end up on the other side of the corridor.
> 

I tested it for this notebook. The shortest path goes in the middle of the `corridor_initial`, not slightly. If we  allow this, I suggest decreasing `buffer_dist <- 500  # distance (in m) from the water stream`, then the calculated corridor would be closer to the water stream. 
 
> > 2. The st_network_paths function requires specifying 'from' and 'to' points. When using the entire network, there are four target points. Still, we need to run the st_network_paths function separately for the upper and lower parts of the network. Otherwise, we cannot create a buffer (region) around the river.
> 
> Yes, we would still need to run the shortest path algorithm twice, once for each side of the river. What I was thinking was:
> 
> * split the initial corridor buffer in two lines, which should represent the initial guesses of the corridor edges on the two sides
> * find the start/end points for each of the side of the river by intersecting the two lines with the `bbox`
> * run the method you have implemented using the full network for each pair of points and using the corresponding line as target geometry (instead of the polygon representing the initial guess of the corridor).
> 
> What do you think? Not needed to be addressed here/now, but I would love to hear your thoughts.

right, we can implement a different approach for finding the intersection points regardless of the method of calculating shortest paths. If it is okay, I submit an issue and will address it in a separate PR.

fnattino reviewed (approved):
Great @SarahAlidoost, thanks a lot! Totally agree in addressing additional steps in a new PR, go ahead and merge for my side!
